various changes

teverett opened the pull request:
A rather large PR that covers many changes

teverett committed:
clang2

teverett committed:
no longer using intel syntax on GCC

teverett committed:
uncomment out stuff

teverett committed:
update to readme

teverett committed:
qemu

teverett committed:
use objcopy

teverett committed:
make games

teverett committed:
PIC

teverett committed:
norelax

teverett committed:
build refactoring

teverett committed:
oops

teverett committed:
TCG

teverett committed:
remove accel for now

teverett committed:
oops

teverett committed:
qemu fixes

teverett committed:
removed unneeded macro

teverett committed:
interrupt routing

added code to nicely route interrupts

teverett committed:
space

teverett committed:
serial

teverett committed:
okee-do

teverett committed:
bridges and ethernets and usb, on my!

teverett committed:
brdge

teverett committed:
spaces

teverett committed:
missed rm of cosmos.elf

teverett committed:
array

teverett committed:
space

teverett committed:
reduce output

teverett committed:
show devs

teverett committed:
serial console

teverett committed:
restore the output i removed

teverett committed:
video message

teverett committed:
idt_init from InitIDT to be consistent with other init functions

teverett committed:
device fun

teverett committed:
ok

teverett committed:
fancy device registration

teverett committed:
device registration

teverett committed:
docs

teverett committed:
display device

can eventually be merged with VGA device

teverett committed:
fixes

teverett committed:
reorder startup

teverett committed:
serial read and echo

teverett committed:
space

teverett committed:
fix

teverett committed:
ugh clang

teverett committed:
build changes

teverett committed:
all branches

teverett committed:
naming

teverett committed:
naming

teverett committed:
switch to apt

teverett committed:
osx

teverett committed:
no clang for now

teverett committed:
packed

teverett committed:
more network devices

teverett committed:
fix multiple inclusion of P macro

teverett committed:
set up irqs

teverett committed:
pci stuff

teverett committed:
brew

teverett committed:
brew

teverett committed:
x86_64-elf-gcc

teverett committed:
fix various ifdefs

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ifdef fixes

teverett committed:
ugh

teverett committed:
PCI registration

teverett committed:
device search

teverett committed:
dev search

teverett committed:
dev search

teverett committed:
convert ATA to use pci_search

teverett committed:
less output

teverett committed:
add strcpy

teverett committed:
device description

teverett committed:
device names

teverett committed:
better rs232 init

teverett committed:
better

teverett committed:
f

teverett committed:
renaming

teverett committed:
space

teverett committed:
docs

teverett committed:
better

teverett committed:
device types

teverett committed:
better

teverett committed:
array protections

teverett committed:
validations ….

teverett committed:
better

teverett committed:
docs

teverett committed:
split timing

timing contiains a device (rtc) and some functionality (sleep).

split out the device.

teverett committed:
refactor to get apis out of headers

teverett committed:
hardening

teverett committed:
docs

teverett committed:
refactoring

teverett committed:
start of dynArray

teverett committed:
add ability to shrink array

teverett committed:
start of list

teverett committed:
lists!

teverett committed:
list

teverett committed:
convert devicemanager to use list

teverett committed:
convert interrupt router to use lists and arrays

teverett committed:
better

teverett committed:
added iterators

teverett committed:
use the iterator

teverett committed:
more rs232 ports

teverett committed:
docs

teverett committed:
pass stack frame in interrupt handler

teverett committed:
add a ringbuffer

teverett committed:
rb

teverett committed:
rb

teverett committed:
pointer size fix

teverett committed:
string

teverett committed:
kernelstring

teverett committed:
length

teverett committed:
k

teverett committed:
kernelstring

teverett committed:
update

teverett committed:
bridge fun

teverett committed:
display name

teverett committed:
ne2000 fun

teverett committed:
more kernelstring

teverett committed:
kernelstring

teverett committed:
oops

teverett committed:
sd

teverett committed:
fixes

teverett committed:
removed unused define

teverett committed:
factor out pic

in advance of making it a standalone device

teverett committed:
pic externalized

teverett committed:
oops

teverett committed:
yup

teverett committed:
ata has more typical init now

teverett committed:
oops

teverett committed:
ATA now uses list

teverett committed:
removed unused variable

teverett committed:
ethernet

teverett committed:
qemu options

teverett committed:
oops

teverett committed:
refinement

teverett committed:
refactoring devices into /dev/

teverett committed:
refactoring devices into /dev

teverett committed:
refactoring devices into /dev

teverett committed:
refactored pic into /dev

teverett committed:
refactored network into /dev

teverett committed:
refactored ata into /dev

teverett committed:
convert pci to use list

teverett committed:
move pci into /dev

teverett committed:
sleeeeeeep

teverett committed:
devs in their own archive

teverett committed:
move the dev startup

teverett committed:
cleaner startup

teverett committed:
floppy

teverett committed:
mouse works!

teverett committed:
mouse fixes

teverett committed:
docs

teverett committed:
start of floppy

teverett committed:
refactor out cmos

its actually needed by the floppy driver too….

teverett committed:
better output

teverett committed:
better panic messages

teverett committed:
ugh, important fix!

teverett committed:
speaker support

teverett committed:
refactor rtc

teverett committed:
list and array refinements

teverett committed:
rtc subscriptions

teverett committed:
oops

teverett committed:
rtc fun

teverett committed:
PIT support

teverett committed:
RTC not PIT

teverett committed:
enable auto dev

teverett committed:
speaker works

teverett committed:
pit fun

teverett committed:
pit works.  ready to connect in context switches

teverett committed:
its sat evening and im bored

teverett committed:
better boot output

teverett committed:
e1000 stub

teverett committed:
can read NE2000 MAC

teverett committed:
well now im exhausted

teverett committed:
sadly, no interrupts are recieved from NE2000

teverett committed:
rename

teverett committed:
ne2000

kurtmweber commented on Makefile:
```diff
@@ -1,20 +1,32 @@
-all: boot-subsystem kernel-subsystem
+BOOTIMAGE=hda.img
 
+all: bootimage
```
Could the creation of a bootable image be moved to a separate target so that it's not part of all?  It's completely a personal preference on my part, but I prefer to keep the disk images and whatnot in a separate tree because to me they're conceptually different parts.  Maybe something like make all-bootimage (or ab to save typing?) so that those who prefer can have a straightforward way to build the image within the build tree, while those who don't don't have it put there.

This also would make it easier for people who want to experiment with different disk sizes (probably not an issue at the moment, but as it grows) to do so without having their changes overwritten every time they pull from the repo (there are other practical issues I can see emerging if the project moves along, with overwriting data, etc.).

teverett replied:
Yup, no problem.

teverett replied:
Done.  Make all now just makes the bin files

kurtmweber commented on kernel/asm/io.c:
```diff
@@ -5,50 +5,47 @@
-
 #include <types.h>
 
-static inline uint8_t asm_in_b(uint16_t port){
```
With optimizations turned off (which I've done to avoid unpredictable complications with debugging), GCC doesn't inline functions unless it's explicitly told to.  I made the IO functions inline because, while it's not an issue yet, the function call overhead may cause performance issues when reading or writing large amounts of data via IO ports, when that's necessary for devices that don't support DMA.

teverett replied:
I'll try making it inline again and see if it works.

kurtmweber replied:
Oh, was inlining causing issues?

teverett replied:
Ya.  I'm trying to fix them  now.   Given that's inlining is a performance issue can we pull this as-is and I'll submit a ticket to fix it?

kurtmweber replied:
Gotcha, that sounds like a plan then.

teverett replied:
Issue added.  Thanks for your flexibility.

kurtmweber commented on kernel/asm/misc.c:
```diff
@@ -31,7 +28,7 @@ pttentry asm_cr3_read(){
 	
 	asm volatile(
-		"mov %0, cr3"
+		"mov %%cr3 ,%0"
```
Once I merge this, I'm going to add some preprocessor directives so that it still uses Intel syntax on build platforms that support it.  It's just a personal preference on my part, and it also keeps it consistent with the bootloader code.  Hypothetically, clang-darwin might eventually support Intel syntax too, and once it does I think all AT&T inline asm should go for consistency's sake.

teverett replied:
This is fine.  I wonder if we could externalize all the inline assembler into ASM files.  That way, the differences in how clang and gcc handle inline assembler wouldn't matter

kurtmweber commented on kernel/collection/array/array.c:
```diff
@@ -0,0 +1,104 @@
+#include <panic/panic.h>
+#include <console/console.h>
+
+struct array* arrayNew(uint32_t size) {
```
Again, another style comment, but for consistency's sake I think we should standardize on a symbol name convention, especially for symbols that are intended to be visible outside the module they're defined in.  Personal preference and an exercise of personal privilege here, but since I already established the practice of using underscores rather than camelcase, that's what I'm going to run with going forward.

teverett replied:
I'll fix them.  I have no issue with underscores.   My brain was in Java mode when I wrote this :)

teverett replied:
lotsa fixes added.

kurtmweber commented on kernel/collection/ringbuffer/ringbuffer.c:
```diff
@@ -0,0 +1,78 @@
+//*****************************************************************
```
Do we need arbitrary-size ring buffers yet?  I wonder if for now, we can use fixed-size (256-element, maybe 65536-element if needed) ringbuffers implemented under the hood as 256- (or 64k-)element arrays, and we can take advantage of unsigned integer wraparound to remove some of the complexity with dealing with the wraparound of indices?

teverett replied:
100% agree.  I can change the ringbuffer to fixed size.

teverett replied:
Issue added for this.  currently ringbuffer isn't actually used, therefore we can make it what we want.

kurtmweber commented on kernel/console/drivers/serial/serial_console.c:
```diff
@@ -0,0 +1,25 @@
+
+void serial_console_register(){
+	console_interfaces[CONSOLE_DRIVER_SERIAL].init = &serial_console_init;
+	console_interfaces[CONSOLE_DRIVER_SERIAL].setpos =0;
```
I'll update console_setpos to add a check to ensure that the setpos function is defined for the driver before calling it, to avoid bad things happening if someone tries to run setpos on a console that doesn't have that implemented.  I should have done it from the start.

teverett replied:
Thanks!

kurtmweber commented on kernel/Makefile:
```diff
@@ -2,98 +2,113 @@ include ../mk/build.mk
-cosmos: cosmos.o asm-dir ata-dir console-dir interrupts-dir keyboard-dir mm-dir panic-dir pci-dir string-dir timing-dir video-dir
-	$(LD) $(LDFLAGS) -o cosmos \
+cosmos: $(KERNEL_ELF)
+	$(OBJCOPY) $(OBJCOPYFLAGS) $(KERNEL_ELF) $(KERNEL_BIN)
```
Is there an advantage to doing it like this?

teverett replied:
yup.  There are debuggers that can read ELF binaries and debug into them.

teverett committed:
fix for code review comment

all no longer builds the boot image

teverett committed:
fixed all sorts of camel-case

teverett committed:
camel case

teverett committed:
camel case

teverett committed:
coding standards

teverett committed:
idt refactoring

teverett committed:
idt fun

teverett committed:
better naming

teverett committed:
better naming

teverett committed:
keyboard refactoring

teverett committed:
just a bit cleaner

teverett committed:
closer….

teverett committed:
len

teverett committed:
string overflow

teverett committed:
nasm invocation changes

teverett committed:
monitor

kurtmweber commented on kernel/console/kprintf.c:
```diff
@@ -45,11 +44,11 @@ uint64_t kprintf(const char *s, ...){
-		if (idx_out == 63){
-			out_str[63] = '\0';
+		if (idx_out == KPRINTF_LEN){
+			out_str[KPRINTF_LEN] = '\0';
```
Will this not cause buffer overrun issues?  `out_str[KPRINTF_LEN]` won't actually exist, the range of indices would be from 0 to `KPRINTF_LEN - 1` (that's why I checked for a maximum index of 63 in my original version).

kurtmweber replied:
I think this will be a buffer overrun--`out_str[KPRINTF_LEN]` won't actually exist, no?

kurtmweber commented on kernel/console/kprintf.c:
```diff
@@ -73,11 +72,11 @@ uint64_t kprintf_do_uint(uint8_t width, char *output, uint8_t base, __builtin_va
 		case 32:
 			conv_uint = __builtin_va_arg(ap, unsigned int);
-			uitoa3(conv_uint, output, 66, base);
+			uitoa3(conv_uint, output, KPRINTF_LEN, base);
```
I can't imagine that there'd be any reason for the # of chars in the output string to be over 66 in `kprintf_do_uint()`.  This function is concerned specifically with stringifying unsigned integer values, so if we do a conversion to a binary representation, then at most we'll have 64 bytes to display the stringified value (for a uint64_t), an optional "b" sigil at the end, and \0.

kurtmweber commented on kernel/console/kprintf.c:
```diff
@@ -73,11 +72,11 @@ uint64_t kprintf_do_uint(uint8_t width, char *output, uint8_t base, __builtin_va
 		case 64:
 			conv_ulong = __builtin_va_arg(ap, unsigned long);
-			uitoa3(conv_ulong, output, 66, base);
+			uitoa3(conv_ulong, output, KPRINTF_LEN, base);
```
See earlier comment re: uint conversions

kurtmweber commented on kernel/console/kprintf.c:
```diff
@@ -89,7 +88,7 @@ uint64_t kprintf_proc_format_string(const char *s, uint64_t *chars_written, __bu
 	// possible "b" sigil at end
 	// NULL terminator
-	char output[66];
+	char output[KPRINTF_LEN];
```
This value is used only to hold the result of a <something-other-than-a-string> conversion.  Strings referenced in the format string are just copied directly, and anything else will have a string representation of no more than 66 bytes.

I see the reason to have a 1024-byte buffer in the outermost `kprintf()` function which handles actually sending the string to the console, since console writes can be slow so this can save some calls there, but I'm not sure there's any gain to using a larger buffer in the conversion-specifier handling code.

teverett replied:
Changes reverted

teverett committed:
reverted changes to kprintf

teverett committed:
start of sb16

teverett committed:
missed fields

teverett committed:
krealloc / kfree workaround

There is somethig weird happening in krealloc and kfree.  this patch works around it, at the cost of leaking some memory.

kurtmweber commented on kernel/collection/array/array.c:
```diff
@@ -0,0 +1,125 @@
+    if (0!=arr){
+        if (size >= arr->size){
+            /*
+            * TODO.  Should use krealloc here, but I think there is a bug in krealloc
```
What are you seeing in krealloc?  I'll take a stab at it, I'm working on MMU stuff anyway.

teverett replied:
if I used krealloc, I get random crashes.  

So, I re-wrote the code to just allocate a new block of memory, copy the data and release the old block: it crashes too.

If you uncomment the line with kfree() on it, you will see that releasing that block of memory causes a crash.

teverett committed:
Merge branch 'master' of https://github.com/kurtmweber/CosmOS into dev

kurtmweber commented on kernel/string/itoa.c:
```diff
@@ -5,12 +5,9 @@
 #include <types.h>
 
-void uitoa3(uint64_t n, char *s, uint8_t len, uint8_t base){
+void uitoa3(uint64_t n, char *s, uint16_t len, uint8_t base){
```
If len is changed from uint8_t to uint16_t, then the index variable (i) for iterating over s needs to be changed accordingly or it won't work.

teverett replied:
fixed, thanks

kurtmweber replied:
Now that I look at it again...there shouldn't be a need for a uint16_t.  This function gets called to convert an integer (n) into a character string (s, which is passed by the caller as a pointer and is effectively the return value).  There is no circumstance in which a 64-bit unsigned integer-to-string conversion would result in a string longer than 65 bytes + \0, so unless there's something I'm not considering it seems like a uint8_t should be sufficient for all cases, no?  Even if the caller passes a pointer to an array longer than 256 bytes, this function will never have any reason to write beyond s[65].

On the other hand, i and len are all on the stack, so they're both essentially taking up 64 bits each regardless under the hood whether they're uint8 or uint16 and it's not actually making the kernel image larger, so it's probably not worth worrying about (unless, again, there's something I'm not thinking of).

teverett committed:
itoa fix

teverett commented:
Thanks!
